presidio-analyzer/analyzer/analyzer_engine.py

```diff
@@ -96,23 +103,161 @@ def analyze(self, text, entities, language, all_fields):
                                  "Either have all_fields set to True "
                                  "and entities are empty, or all_fields "
                                  "is False and entities is populated")
-            # Since all_fields=True, list all entities by going
+            # Since all_fields=True, list all entities by iterating
             # over all recognizers
             entities = self.__list_entities(recognizers)
 
+        # run the nlp pipeline over the given text, store the results in
+        # a NlpArtifacts instance
+        nlp_artifacts = self.nlp_engine.process_text(text, language)
         results = []
         for recognizer in recognizers:
             # Lazy loading of the relevant recognizers
             if not recognizer.is_loaded:
                 recognizer.load()
                 recognizer.is_loaded = True
 
-            r = recognizer.analyze(text, entities)
-            if r is not None:
-                results.extend(r)
+            raw_results = recognizer.analyze(text, entities, nlp_artifacts)
+            if raw_results is not None:
+                # try to improve the results score using the surronding context
+                # words
+                raw_results = \
+                    self.enhance_using_context(
```

Comment:
Now that we don't really fuse in Spacy into each recognizer, it makes sense to have the context handling within the PatternRecognizer logic, or create a new class for context handling. I'm not sure the AnalyzerEngine is the right place to handle context. Should be just the orchestrator

Reply:
discussed f2f. moved into the EntityRecognizer, where it can be used later by the PatternRecognizers and even new types.
Please note I refactored and extracted the relevant tests from the analyzer engine into a new file